### PR TITLE
Compile with GHC8.10.2

### PR DESCRIPTION
**Pull request by nc6**
This PR does not yet update CI to build with GHC 8.10, but makes various changes that bring the code into a place where it compiles with GHC 8.10. As such, downstream libraries may choose to build it with later GHC versions, and it should be easier to move to 8.10 when we decide to make the switch.

### CODE REVIEW

**Comment by mrBliss** [comment]
It seems that the build failed while compiling GHC itself on `x86_64-w64-mingw32` :slightly_frowning_face:

**Comment by nc6** [comment]
And a strange error about no `HasResolution` in the cardano-binary tests.

File: binary/cardano-binary.cabal [hunk 1]
```diff
@@ -60,7 +60,12 @@ library
                        -Wredundant-constraints
                        -Wpartial-fields
                        -Wcompat
-
+  if impl(ghc >=8.8)
+    ghc-options:       -fno-warn-missing-deriving-strategies
+  if impl(ghc >=8.10)
+    ghc-options:       -fno-warn-missing-safe-haskell-mode
+                       -fno-warn-prepositive-qualified-module
+                       -fno-warn-unused-packages
```
Comment (mrBliss) [inline binary/cardano-binary.cabal:68]: I prefer `-Wno-X` over `-fno-warn-X`, then it's consistent with the warning flags we use above.

Comment (nc6) [reply binary/cardano-binary.cabal:68]: I have no opinion on this; I'm just taking the same approach as was used in `cardano-prelude`https://github.com/input-output-hk/cardano-prelude/blob/master/cardano-prelude.cabal#L79

Comment (mrBliss) [reply binary/cardano-binary.cabal:68]: Yeah, I saw that too. Would you mind if I pushed a commit to this PR that changed it to `-Wno-X`?

Comment (nc6) [reply binary/cardano-binary.cabal:68]: Not at all!

Comment (mrBliss) [reply binary/cardano-binary.cabal]: Done

File: binary/cardano-binary.cabal [hunk 2]
```diff
@@ -60,7 +60,12 @@ library
                        -Wredundant-constraints
                        -Wpartial-fields
                        -Wcompat
-
+  if impl(ghc >=8.8)
+    ghc-options:       -Wno-missing-deriving-strategies
+  if impl(ghc >=8.10)
+    ghc-options:       -Wno-missing-safe-haskell-mode
+                       -Wno-prepositive-qualified-module
+                       -Wno-unused-packages
```
Comment (mrBliss) [inline binary/cardano-binary.cabal:68]: Actually, are all these flags that disable warnings even needed (in the other cabal files as well)? I assume you have copied them from `cardano-prelude`, but there `-Weverything` is used. Here `-Wall` is used, so these extra warnings don't have to be disabled. (I didn't have to disable them in https://github.com/input-output-hk/ouroboros-network/pull/2542)

I prefer `-Wall` + a manual selection of flags over `-Weverything`, which forces you to disable some flags based on the GHC version.

**Review by mrBliss — APPROVED** [review]
LGTM, I think you can remove these conditionals from the cabal files.

File: binary/cardano-binary.cabal (hunk 2 above)
Comment (nc6) [reply binary/cardano-binary.cabal:68]: Hmm, certainly some of them were needed in some places. But it's true that after adding them to a couple of files, I pre-emptively added them in other places...

Comment (mrBliss) [reply binary/cardano-binary.cabal:68]: BTW, I wouldn't disable `-Wno-unused-packages`, I'd just fix the warnings :slightly_smiling_face:

Comment (mrBliss) [reply binary/cardano-binary.cabal:68]: I just tried removing all those flags myself and it compiled without warnings (except for those from the C code).

Comment (mrBliss) [reply binary/cardano-binary.cabal:68]: > I prefer `-Wall` + a manual selection of flags over `-Weverything`, which forces you to disable some flags based on the GHC version.

In other words, no changes are needed to the cabal files at all.

**Comment by mrBliss** [comment]
I'm assuming you're planning on rebasing/squashing? It would be nice to get those two merge commits out of the history.

**Pull request merged**